crm_timesheet/models/crm_phonecall.py

```diff
@@ -0,0 +1,90 @@
+##############################################################################
+
+from openerp import models, fields, api
+from openerp.fields import DATE_LENGTH
+from datetime import datetime
+from openerp.tools.translate import _
+
+
+class CrmPhonecall(models.Model):
+    _inherit = 'crm.phonecall'
+
+    analytic_account = fields.Many2one(
+        comodel_name='account.analytic.account')
+    timesheets = fields.One2many(comodel_name='hr.analytic.timesheet',
+                                 inverse_name='phonecall')
+
+    def _timesheet_prepare(self, vals):
+        """
+            NOTE: Call this method ONLY from api.model or api.one methods
+
+            param: vals
+            type: dict
+            desc: crm.phonecall vals dict (It's like create and write methods)
+        """
+        def value_or_id(val):
+            return val if isinstance(val, (bool, int, long)) else val.id
+        date = vals.get('date', getattr(self, 'date', False))
+        user_id = vals.get('user_id', getattr(self, 'user_id', False))
+        account_id = vals.get('analytic_account', self.analytic_account)
+        unit_amount = vals.get('duration', self.duration)
+        res = {
+            'date': date[:DATE_LENGTH] if date else False,
+            'user_id': value_or_id(user_id),
+            'name': vals.get('name', self.name),
+            'account_id': value_or_id(account_id),
+            'unit_amount': unit_amount / 60.0,
+            'code': 'phone',
+            'journal_id': self.env[
+                'hr.analytic.timesheet']._getAnalyticJournal()
+        }
+        return res
+
+    @api.model
+    def create(self, vals):
+        if vals.get('analytic_account') and vals.get('duration', 0) > 0:
+            if not vals.get('date'):
+                raise Warning(_('Date field must be filled.'))
+            timesheet_data = self._timesheet_prepare(vals)
+            vals['timesheets'] = vals.get('timesheets', [])
+            vals['timesheets'].append((0, 0, timesheet_data))
+        res = super(CrmPhonecall, self).create(vals)
+        return res
+
+    @api.one
+    def write(self, vals):
+        timesheet = self.env['hr.analytic.timesheet'].search([
+            ('code', '=', 'phone'), ('phonecall', '=', self.id)])
+        analytic_account = vals.get('analytic_account', self.analytic_account)
+        duration = vals.get('duration', 0)
+        if timesheet:
+            if 'analytic_account' in vals and not vals['analytic_account']:
+                vals['timesheets'] = [(2, timesheet.id, 0)]
+            else:
+                vals['timesheets'] = [(1, timesheet.id,
+                                       self._timesheet_prepare(vals))]
+        elif analytic_account and duration > 0:
+            vals['timesheets'] = [(0, 0, self._timesheet_prepare(vals))]
+        res = super(CrmPhonecall, self).write(vals)
+        return res
+
+    @api.multi
+    def button_end_call(self, context=None, end_date=None):
```

Comment:
context is not needed when you decorate a method with `@api.multi`. Context is accessed via `self.env.context` and is always initialized (so no more `if not context...`)